compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1848,9 +1830,16 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             .map(|(trait_def_id, mut impls)| {
                 // Bring everything into deterministic order for hashing
                 impls.sort_by_cached_key(|&(index, _)| {
```

Comment:
`impls_in_crate` returns a `FxIndexMap<DefId, Vec<(DefId, Option<SimplifiedType>)>>`; The `Vec` part should be in source code order, and the `Map<TraitId, ...>` part is ordered by the source code order of the Impls. Since the source code is hashed into crate_hash, I believe you're right, we can remove both of these sorts. 

In fact, if I understand correctly, we can also remove a third sort in `encode_incoherent_impls`, since it ultimately relies on a similarly stable iteration through `tcx.hir().items()`.